airbyte-ci/connectors/regression-testing/src/regression_testing/connector_runner.py

```diff
@@ -0,0 +1,378 @@
+        return await self._connector_under_test_container.label(label)
+
+    async def get_container_entrypoint(self):
+        entrypoint = await self._connector_under_test_container.entrypoint()
+        return " ".join(entrypoint)
+
+    async def _run(
+        self,
+        airbyte_command: List[str],
+        raise_container_error: bool,
+        config: SecretDict = None,
+        catalog: dict = None,
+        state: Union[dict, list] = None,
+        enable_caching=True,
+    ) -> List[AirbyteMessage]:
+        """Run a command in the connector container and return the list of AirbyteMessages emitted by the connector.
+
+        Args:
+            airbyte_command (List[str]): The command to run in the connector container.
+            raise_container_error (bool): Whether to raise an error if the container fails to run the command.
+            config (SecretDict, optional): The config to mount to the container. Defaults to None.
+            catalog (dict, optional): The catalog to mount to the container. Defaults to None.
+            state (Union[dict, list], optional): The state to mount to the container. Defaults to None.
+            enable_caching (bool, optional): Whether to enable command output caching. Defaults to True.
+
+        Returns:
+            List[AirbyteMessage]: The list of AirbyteMessages emitted by the connector.
+        """
+        container = self._connector_under_test_container
+        if not enable_caching:
+            container = container.with_env_variable("CAT_CACHEBUSTER", str(uuid.uuid4()))
+        if config:
+            container = container.with_new_file(self.IN_CONTAINER_CONFIG_PATH, contents=json.dumps(dict(config)))
+        if state:
+            container = container.with_new_file(self.IN_CONTAINER_STATE_PATH, contents=json.dumps(state))
+        if catalog:
+            container = container.with_new_file(self.IN_CONTAINER_CATALOG_PATH, contents=catalog.json())
+
+        await self._write_output(airbyte_command, container)
+
+    async def _write_output(self, airbyte_command: list, container: dagger.Container):
+        filename = f"raw_output.txt"
+        local_output_file_path = f"{self._output_directory}/{filename}"
+
+        entrypoint = await container.entrypoint()
+        airbyte_command = entrypoint + airbyte_command
+        container = container.with_exec(
+            ["sh", "-c", " ".join(airbyte_command) + f" > {self.IN_CONTAINER_OUTPUT_PATH} 2>&1 | tee -a {self.IN_CONTAINER_OUTPUT_PATH}"],
+            skip_entrypoint=True,
+        )
+        await container.file(self.IN_CONTAINER_OUTPUT_PATH).export(local_output_file_path)
+        await self._write_comparable_records_and_state(local_output_file_path)
+
+    async def _write_comparable_records_and_state(self, filepath: str):
+        raw_output = await AnyioPath(filepath).read_text()
+        await self._backend.write(self._raw_output_iter(raw_output))
+
+    def _raw_output_iter(self, raw_output):
```

Comment:
I think the runner should not perform any deserialization. I'd suggest deserialization to happen post `dispatch` when we actually need deserialized stuffed in tests or diff.
My point is that error detection should happen at a more downstream level where actual good or bad connector behavior will be evaluated.

I like the flow of:
1. run commands on containers and gather raw outputs (stdout, stderr)
2. build objects from these output with backends. 
3. Use a backend useful for test, another one for debugging, etc.